File: linked-list/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // Two pointers
        // Floyd's Tortios & Hare 
        ListNode fast = head;
        ListNode slow = head;

        while (fast != null && fast.next != null) {
            if (fast == slow) {
                return true;
            } else {
                fast = fast.next;
                slow = slow.next.next;
            }
        }

        return false;
    }
}

File: linked-list/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        cur = head
        
        while cur and cur.next and cur.next.next:
            slow, fast = cur, cur
            if slow != fast:
                slow = slow.next
                fast = fast.next.next
            else:
                return True
        
        return False